# Tutorial: Shotgun metagenomics with MOSHPIT

This repository contains materials accompanying the `<NAME OF THE WORKSHOP>`. To jump directly to the analysis notebooks, click on the badges for the corresponding tutorials below.

The notebook included in this repository describes simplified analyses of microbiome datasets using the with the MOSHPIT toolkit build using the [QIIME 2](https://qiime2.org/) framework. To learn more about the QIIME 2 project and applications in microbiome research, visit https://qiime2.org/. These materials have been adapted from the [**Advanced Block Course: Computational Biology**](https://github.com/bokulich-lab/advanced-comp-bio-tutorial.git); all source code is licensed under the Apache License 2.0.

## Functional analysis with MOSHPIT
[![Workshop Notebook (Colab)](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/bokulich-lab/rigi-workshop/blob/main/notebook.ipynb)

In this notebook, we use feature tables generated for MAGs obtained from analysis of coca beans fermentation (see the original study [here](https://www.sciencedirect.com/science/article/pii/S0740002020301970)). To launch, click on the badge above.

## Source attribution

This analysis notebook builds on top of the [MOSHPIT tutorial](https://moshpit.readthedocs.io).

The Google Colab setup script was adapted from the [**Advanced Block Course: Computational Biology**](https://github.com/bokulich-lab/advanced-comp-bio-tutorial.git).

## Licensing

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)<br>
All source code is licensed under the Apache License 2.0.
